### PR TITLE
[15.0][MIG] excel_import_export, excel_import_export_demo: migration

### PR DESCRIPTION
**Pull request by Mantux11**
Migration to version `15.0`. Changes by pre-commit and changes because of Odoo code changes.

### CODE REVIEW

**Comment by Mantux11** [comment]
@kittiu Hey. Because of your migration to v13 and v14, I want some help from you. I think, you know how JS worked here and mostly all functionality what JS code does here. Personally, I don't know probably anything about JS, and I saw that Odoo deleted `ActionManager` and I suggest that everything was moved to services or somewhere else. So my questions are:

1. How JS code worked here, what was functionality of JS code, why it was used here?
2. Maybe you could suggest, where I could try to look to understand what is going on here? Maybe you know where everything was moved, what was changed and where I could insert these functions? Some files to look, classes and etc. 

Thank you! :wink:

**Comment by kittiu** [comment]
@Mantux11 you are wrong that I know what I was doing about JS :laughing: 
I just try to imitate the work from other module, i..e, report_csv, report_xlsx.
Now, I saw this already migrated to v15 https://github.com/OCA/reporting-engine/blob/15.0/report_xlsx/static/src/js/report/action_manager_report.esm.js

:)

**Comment by Mantux11** [comment]
@kittiu Thanks for your answer. I will try to change what it's needed to be changed. :D

**Comment by kittiu** [comment]
Not sure this is ready for review. Functional test couple of action without success.

1. Sales > Reporting > Partner List Report
image
2. Sales > Sales Order > Quotation / Order (.xlsx) ----> not in the print action yet.
3. Technical > Excel Import/Export > Sample Lead Report, error when click to export.
image

**Comment by Mantux11** [comment]
> Not sure this is ready for review. Functional test couple of action without success.
> 
> 
> 
> 1. Sales > Reporting > Partner List Report
> 
> image
> 
> 2. Sales > Sales Order > Quotation / Order (.xlsx) ----> not in the print action yet.
> 
> 3. Technical > Excel Import/Export > Sample Lead Report, error when click to export.
> 
> image
> 
> 
> 
> 



Okey, I will test more on that. Thanks, for review

**Comment by Mantux11** [comment]
@kittiu Talking about Sample Lead Report, you also missed that, this isn't working in 14 version, because `planned_revenue` field is deleted from 14 version. So I changed to `expected_revenue`, why I changed to that, you could see comparing 13 and 14 `crm.lead` core code. Partner List report is fixed. Talking about SO report, I see that when you press Export XLSX, you could choose what template to use in wizard. Maybe I'm wrong, but this should work as it is? Or there should be one more report? It would be really appreciated, If you could give me more example's on SO report, how it should look like,.

**Comment by gaikaz** [comment]
@OCA/server-environment-maintainers 
Sorry to be pinging, but this is a valid migration that is just not getting any attention for 3 months now. Wouldn't want to be considered stale and closed.
Could we please get a maintainer review or two? 🙏🏻

**Comment by kittiu** [comment]
@gaikaz have you tested this? Can you approve this? Thanks!

**Comment by kittiu** [comment]
Ping @Saran440 @ps-tubtim

**Comment by gaikaz** [comment]
@kittiu 
We'll be migrating one of our modules that depend on this, hopefully over next week. 
Will do my review too.

File: excel_import_export/static/src/js/report/action_manager_report.esm.js
Comment (Du-ma) [inline excel_import_export/static/src/js/report/action_manager_report.esm.js]:
```suggestion
function getReportUrl({report_name, context, data}, env) {
    // Rough copy of action_service.js _getReportUrl method.
    let url = `/report/excel/${report_name}`;
    const actionContext = context || {};
    if (data && JSON.stringify(data) !== "{}") {
        const encodedOptions = encodeURIComponent(JSON.stringify(data));
        const encodedContext = encodeURIComponent(JSON.stringify(actionContext));
        return `${url}?options=${encodedOptions}&context=${encodedContext}`;
    }
    if (actionContext.active_ids) {
        url += `/${actionContext.active_ids.join(",")}`;
    }
    const userContext = encodeURIComponent(JSON.stringify(env.services.user.context));
    return `${url}?context=${userContext}`;
}

async function triggerDownload(action, {onClose}, env) {
    // Rough copy of action_service.js _triggerDownload method.
    env.services.ui.block();
    try {
        await download({
            url: "/report/download",
            data: {
                data: JSON.stringify([getReportUrl(action, env), "excel"]),
                context: JSON.stringify(env.services.user.context),
            },
        });
    } finally {
        env.services.ui.unblock();
    }
    if (action.close_on_report_download) {
        return env.services.action.doAction(
            {type: "ir.actions.act_window_close"},
            {onClose}
        )
    }
    if (onClose) {
        onClose();
    }
}

registry
    .category("ir.actions.report handlers")
    .add("xlsx_handler", async function (action, options, env) {
        if (action.report_type === "excel") {
            await triggerDownload(action, options, env);
            return true;
        }
        return false;
    });
```
How about something like this?
Removed the `console.log`s as well as made the code easier to read (in my opinion).
Hard-coded in some elements specific to this report, for example - How would we have a different type of report rather than `excel`, when it's in the if block?

Comment (Mantux11) [reply excel_import_export/static/src/js/report/action_manager_report.esm.js]: Thanks @Du-ma. I will test that ASAP.

Comment (Du-ma) [reply excel_import_export/static/src/js/report/action_manager_report.esm.js]: Whoops, forgot something, edited my suggestion

**Review by Saran440 — APPROVED** [review]
Test with demo :+1:

**Comment by GreyShim** [comment]
Hello guys,

Is it possible to merge the pull request ? I need this module for a project

Regards

**Comment by kittiu** [comment]
@Mantux11 looks like pre-commit is failing. Can you check please?

**Comment by GreyShim** [comment]
@Mantux11  

do not forget to integrate this change : https://github.com/OCA/server-tools/pull/2283/files
it's very important, data are deleted without this

**Comment by Mantux11** [comment]
@GreyShim I would wait for that PR to be merged. And I would include that here.

**Comment by Mantux11** [comment]
@kittiu, @GreyShim, @Du-ma Could you review it now, please? This module is waiting in the line a long time ago. I changed what @GreyShim said that, and also, included what @luc-demeyer suggested in that PR above.

**Review by norlinhenrik — APPROVED** [review]
Thank you for migrating and fixing this module!

models/common.py
`line = "{}.{}".format("xls", uuid.uuid4()) + "," + line`
To be consistent, please change to
`line = "{}.{}".format("--excel_import_export--", uuid.uuid4()) + "," + line`
with underscore instead of hyphen. (I cannot write underscore since it is evaluated as bold.)

**Comment by gaikaz** [comment]
@Mantux11 Henrik here has very good suggestion. 

@norlinhenrik 
You can do `\_` and have \_\_as many underscores as you want\_\_😉

**Comment by Du-ma** [comment]
Pls don't copy paste that

**Comment by Mantux11** [comment]
@norlinhenrik Changed that. @Du-ma Yes I know, how you like to use `f"{}"` string :D

File: excel_import_export/models/ir_report.py
Comment (Du-ma) [inline excel_import_export/models/ir_report.py]:
```suggestion
        return report_obj.with_context(**context).search(conditions, limit=1)
```
Why do we even need this `context`???
Don't we already have one?
Are we missing something?

File: excel_import_export/wizard/import_xlsx_wizard.py
Comment (Du-ma) [inline excel_import_export/wizard/import_xlsx_wizard.py]:
```suggestion
```

File: excel_import_export/wizard/import_xlsx_wizard.py
Comment (Du-ma) [inline excel_import_export/wizard/import_xlsx_wizard.py]:
```suggestion
```

File: excel_import_export/wizard/import_xlsx_wizard.py
Comment (Du-ma) [inline excel_import_export/wizard/import_xlsx_wizard.py]:
```suggestion
                    _('File "%(fname)s" not found in template, %(name)s.', fname=template.fname, name=template.name),
```
For future reference.

**Review by Du-ma — CHANGES_REQUESTED** [review]

**Comment by Du-ma** [comment]
Also a general question, why is this module not in `oca/reporting-engine`?

File: excel_import_export/__manifest__.py
Comment (AungKoKoLin1997) [inline excel_import_export/__manifest__.py:13]:
```suggestion
    "external_dependencies": {"python": ["openpyxl"]},
```

**Review by AungKoKoLin1997 — COMMENTED** [review]
Test with demo data: work as expected
For the functional, there is a minor things.

We can remove `xlrd` and `xlwt` from external dependencies because these two are part of base odoo requirements.

Sales > Sales Order > Quotation / Order (.xlsx) ----> not in the print action yet.
We need to click manually to add this report under print action in report definition. But I think this is not a problem. We can export under action.

Screenshot from 2022-12-14 17-32-08

File: excel_import_export/models/ir_report.py
Comment (norlinhenrik) [reply excel_import_export/models/ir_report.py]: I think the context may be removed.

**Comment by norlinhenrik** [comment]
> Also a general question, why is this module not in `oca/reporting-engine`?

I use this module to import data. I prefer keeping it in `oca/server-tools`.

**Comment by kittiu** [comment]
> Also a general question, why is this module not in `oca/reporting-engine`?

This was discussed before, for the fact that it can import too.l, and many time used as data import/export. So being just report does not comply. It is more like tools to build things.

**Comment by Mantux11** [comment]
@Du-ma Included your suggestions. @AungKoKoLin1997 Also deleted dependencies. And about action, so if it's not a problem, I have to add somewhere this action, or leave it as it is?

**Review by Du-ma — APPROVED** [review]

File: excel_import_export_demo/report_action/sale_order/report.xml
Comment (AungKoKoLin1997) [inline excel_import_export_demo/report_action/sale_order/report.xml:9]:
```suggestion
        <field name="report_file">sale_order</field>
        <field name="binding_model_id" ref="sale.model_sale_order"/>
        <field name="binding_type">report</field>
```

**Review by AungKoKoLin1997 — CHANGES_REQUESTED** [review]

File: excel_import_export_demo/report_action/sale_order/report.xml
Comment (AungKoKoLin1997) [reply excel_import_export_demo/report_action/sale_order/report.xml:9]: Please add these lines for 
Sales > Sales Order > Quotation / Order (.xlsx) ----> not in the print action yet.

Comment (Mantux11) [reply excel_import_export_demo/report_action/sale_order/report.xml:9]: Added. Thanks :)

**Review by AungKoKoLin1997 — APPROVED** [review]

**Comment by AungKoKoLin1997** [comment]
> We can remove `xlrd` and `xlwt` from external dependencies because these two are part of base odoo requirements.

@Mantux11 Sorry. I didn't see this PR in 14.0. Can you please pay attention to this PR https://github.com/OCA/server-tools/pull/2349 and add in your commit history?

**Comment by Mantux11** [comment]
@AungKoKoLin1997 I included all commit's from `14.0`, Could you review, and maybe merge this?

**Comment by Mantux11** [comment]
@AungKoKoLin1997 I included all newest commit's history from `14.0` with commit that removes mentioned libaries, Could you review, and maybe merge this?

Or @kittiu Could you merge?

**Review by gaikaz — APPROVED** [review]

**Comment by gaikaz** [comment]
@kittiu 
I can finally approve this 🎉 
We have a client with a big excel report that uses this module for generating it and after we now have migrated the client that uses this module while internally fixing any found issues with @Mantux11, it now definitely works.

**Comment by kittiu** [comment]
Thank you all but I don't have power to merge. @dreispt can you help merge this? Thanks!

**Comment by gaikaz** [comment]
@OCA/server-environment-maintainers 
Pretty please, review this 🙏🏻 This has a lot of approves with functional tests as well. 
If we (Via laurea) find any issues, we'll make a PR to fix. Pinky promise.

**Pull request merged**
